### PR TITLE
Fix: ref type in Animated Component

### LINKED ISSUE #42806: [react-native] New AnimatedComponent type doesn't accept refs
The AnimatedComponent<T> type added in change https://github.com/DefinitelyTyped/DefinitelyTyped/commit/f7ba6e840d343675432dfd7886d112977d1531a0 are most welcome, but currently don't accept a `ref` prop. This is especially problematic on ScrollViews and such, as being able to call the various imperative `scrollTo` methods is pretty important.

```
Property 'ref' does not exist on type 'IntrinsicAttributes & AnimatedProps<ScrollViewProps> & { children?: ReactNode; }'.ts(2322)
```

- [x] I tried using the `@types/xxxx` package and had problems.
- [x] I tried using the latest stable version of tsc. https://www.npmjs.com/package/typescript
- [x] I have a question that is inappropriate for [StackOverflow](https://stackoverflow.com/).  (Please ask any appropriate questions there).
- [x] [Mention](https://github.com/blog/821-mention-somebody-they-re-notified) the authors (see `Definitions by:` in `index.d.ts`) so they can respond.
  - Authors: @alloy @zusinShinpei

### PR DESCRIPTION
**Pull request by alloy**
Fixes https://github.com/DefinitelyTyped/DefinitelyTyped/issues/42806

These are the changes from https://github.com/DefinitelyTyped/DefinitelyTyped/pull/42446 re-applied on top of current `master`.

@zusinShinpei I don’t think these changes are correct at the moment, though. What I understand from [the upstream change](https://github.com/facebook/react-native/commit/66e72bb4e00aafbcb9f450ed5db261d98f99f82a) is that the `ref` of an animated component should return the _underlying_ component, i.e. the non-animated one.

The test coverage in the upstream change is a little hard to understand, but this is what I understand it should do:

```tsx
<Animated.View ref={ref => assert(ref instanceof View)} />
```

### CODE REVIEW

File: types/react-native/test/animated.tsx [hunk 1]
```diff
@@ -91,14 +131,17 @@ function TestAnimatedAPI() {
     return (
         <View>
             <Animated.View
+                ref={AnimatedViewRef}
```
Comment (alloy) [inline types/react-native/test/animated.tsx]: `AnimatedViewRef` is defined above as:

```ts
const AnimatedViewRef = React.useRef<typeof Animated.View>(null);
```

…but I believe that should be:

```ts
const AnimatedViewRef = React.useRef<typeof View>(null);
```

**Comment by alloy** [comment]
cc @TheSavior

**Comment by TheSavior** [comment]
Yeah, Animated uses forwardRef now but still exposes getNode for backwards compatibility. The TypeScript changes look a bit too complex for me to confidently review, can someone else review?

**Comment by alloy** [comment]
> Yeah, Animated uses forwardRef now but still exposes getNode for backwards compatibility. The TypeScript changes look a bit too complex for me to confidently review, can someone else review?

@TheSavior 👍 The one thing I need to know for sure is whether or not my understanding is correct, in that the ref of `Animated.View` is an instance of `View`?

**Comment by TheSavior** [comment]
Yeah, the way to think of the typing is that a ref to createAnimatedComponent(MyComponent) gives a ref to MyComponent. Animated.View and others just use createAnimatedComponent.

**Comment by alloy** [comment]
> Yeah, the way to think of the typing is that a ref to createAnimatedComponent(MyComponent) gives a ref to MyComponent. Animated.View and others just use createAnimatedComponent.

Thanks 👍 I just pushed a change to the test of how I think it should work.

**Review by elicwhite — COMMENTED** [review]
I think the tests look right to me

**Comment by alloy** [comment]
@typescript-bot You do you, buddy

**Pull request closed without merging**